src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -238,33 +252,36 @@ private void AddCompletionItems(List<MatchResult> list, CancellationToken cancel
                 // FilterStateHelper is used to decide whether a given item should be included in the list based on the state of filter/expander buttons.
                 var filterHelper = new FilterStateHelper(_snapshotData.SelectedFilters);
 
-                using var _1 = PooledHashSet<string>.GetInstance(out var includedPreferredItems);
-                using var _2 = ArrayBuilder<MatchResult>.GetInstance(out var includedDefaults);
-                var unmatchedDefaults = _snapshotData.Defaults;
+                var includedPreferredItems = new ConcurrentSet<string>();
+                var includedDefaults = new ConcurrentDictionary<string, MatchResult>();
+
+                Enumerable.Range(0, _snapshotData.InitialSortedItemList.Count)
+                    .AsParallel()
+                    .WithCancellation(cancellationToken)
+                    .ForAll(CreateMatchResultAndProcessMatchingDefaults);
 
-                // Filter items based on the selected filters and matching.
-                var totalCount = _snapshotData.InitialSortedItemList.Count;
-                for (var currentIndex = 0; currentIndex < totalCount; currentIndex++)
+                PromoteDefaultItemsToPreferredState();
+
+                void CreateMatchResultAndProcessMatchingDefaults(int index)
                 {
-                    cancellationToken.ThrowIfCancellationRequested();
```

Comment:
i would not mind in a followup pr if we basically tried to be much clearer about setting up parallel state, and what that state mutated.  e.g. by extracting a helper function and passing in things like the collections it is expected to mutate.  it's pretty hard to reason about things here given how complex this is and how much state is implicitly captured.